File: windows/src/AINotebook.Core/Rag/Retriever.cs
using AINotebook.Core.Models;
using AINotebook.Core.Ollama;
using AINotebook.Core.Storage;
using Microsoft.Data.Sqlite;

namespace AINotebook.Core.Rag;

public sealed class Retriever
{
    private readonly NotebookStore _store;
    private readonly IEmbeddingProducing _client;
    public string Model { get; }
    public int RrfK { get; }

    public Retriever(NotebookStore store, IEmbeddingProducing client, string model, int rrfK = 60)
    {
        _store = store;
        _client = client;
        Model = model;
        RrfK = rrfK;
    }

    public async Task<IReadOnlyList<RetrievalHit>> SearchAsync(
        long notebookId, string query, int topK = 8, CancellationToken ct = default)
    {
        // 1) Vector ranking — embed query, brute-force cosine over all embeddings.
        var queryVectors = await _client.EmbedAsync(Model, new[] { query }, ct);
        var queryVector = queryVectors.Length > 0 ? queryVectors[0] : Array.Empty<float>();

        var allEmbeddings = _store.Embeddings(notebookId, Model);
        var vectorRanked = allEmbeddings
            .Select(e => (e.ChunkId, e.SourceId, Score: Cosine.Similarity(queryVector, e.Vector.Values)))
            .OrderByDescending(x => x.Score)
            .Take(topK)
            .ToList();

        // 2) FTS ranking — BM25 top-K within the notebook.
        var ftsRanked = FtsTopK(notebookId, query, topK);

        // 3) Reciprocal Rank Fusion over BOTH lists.
        var rrfScores = new Dictionary<long, float>();
        var meta = new Dictionary<long, (long SourceId, string Snippet)>();
        for (var rank = 0; rank < vectorRanked.Count; rank++)
        {
            var hit = vectorRanked[rank];
            rrfScores[hit.ChunkId] = rrfScores.GetValueOrDefault(hit.ChunkId) + 1.0f / (RrfK + rank + 1);
            meta[hit.ChunkId] = (hit.SourceId, "");
        }
        for (var rank = 0; rank < ftsRanked.Count; rank++)
        {
            var hit = ftsRanked[rank];
            rrfScores[hit.ChunkId] = rrfScores.GetValueOrDefault(hit.ChunkId) + 1.0f / (RrfK + rank + 1);
            meta[hit.ChunkId] = (hit.SourceId, hit.Snippet);  // FTS overwrites with bm25 snippet
        }

        // 4) Hydrate snippets for chunks that only came from the vector branch.
        var missing = meta.Where(kv => kv.Value.Snippet.Length == 0).Select(kv => kv.Key).ToList();
        if (missing.Count > 0)
        {
            var snippets = Snippets(missing);
            foreach (var (id, snip) in snippets)
                meta[id] = (meta[id].SourceId, snip);
        }

        return rrfScores
            .OrderByDescending(kv => kv.Value)
            .Take(topK)
            .Where(kv => meta.ContainsKey(kv.Key))
            .Select(kv => new RetrievalHit(kv.Key, meta[kv.Key].SourceId, kv.Value, meta[kv.Key].Snippet))
            .ToList();
    }

    private List<(long ChunkId, long SourceId, string Snippet)> FtsTopK(long notebookId, string query, int k)
    {
        var conn = _store.Connection;
        using var cmd = conn.CreateCommand();
        cmd.CommandText = """
            SELECT sc.id AS chunk_id, sc.source_id AS source_id, sc.text AS text
            FROM chunks_fts f
            JOIN source_chunks sc ON sc.id = f.chunk_id
            JOIN sources s ON s.id = sc.source_id
            WHERE f.text MATCH $q AND s.notebook_id = $nb
            ORDER BY bm25(chunks_fts)
            LIMIT $k
            """;
        cmd.Parameters.AddWithValue("$q", EscapeFts(query));
        cmd.Parameters.AddWithValue("$nb", notebookId);
        cmd.Parameters.AddWithValue("$k", k);

        var rows = new List<(long, long, string)>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            var text = reader.GetString(2);
            rows.Add((reader.GetInt64(0), reader.GetInt64(1), Prefix240(text)));
        }
        return rows;
    }

    private Dictionary<long, string> Snippets(IReadOnlyList<long> chunkIds)
    {
        var result = new Dictionary<long, string>();
        if (chunkIds.Count == 0) return result;

        var conn = _store.Connection;
        var placeholders = string.Join(",", chunkIds.Select((_, i) => "$p" + i));
        using var cmd = conn.CreateCommand();
        cmd.CommandText = $"SELECT id, text FROM source_chunks WHERE id IN ({placeholders})";
        for (var i = 0; i < chunkIds.Count; i++)
            cmd.Parameters.AddWithValue("$p" + i, chunkIds[i]);

        using var reader = cmd.ExecuteReader();
        while (reader.Read())
            result[reader.GetInt64(0)] = Prefix240(reader.GetString(1));
        return result;
    }

    private static string Prefix240(string text) => text.Length <= 240 ? text : text[..240];

    /// Wrap whole query as an FTS5 phrase: replace " with "" then surround in quotes.
    private static string EscapeFts(string raw) => "\"" + raw.Replace("\"", "\"\"") + "\"";
}

File: Sources/AINotebookCore/Retriever.swift
import Foundation
import GRDB

public actor Retriever {
    private let store: NotebookStore
    private let client: EmbeddingProducing
    private let modelKey: @Sendable () -> String
    public let rrfK: Int

    /// `modelKey` is read at the start of every `search`, so a provider/model
    /// switch in Settings applies to the next query without rebuilding the
    /// retriever (fixes the pre-registry staleness bug).
    public init(
        store: NotebookStore,
        client: EmbeddingProducing,
        modelKey: @escaping @Sendable () -> String,
        rrfK: Int = 60
    ) {
        self.store = store
        self.client = client
        self.modelKey = modelKey
        self.rrfK = rrfK
    }

    /// Convenience for a fixed key (tests, single-provider setups).
    public init(
        store: NotebookStore,
        client: EmbeddingProducing,
        model: String,
        rrfK: Int = 60
    ) {
        self.init(store: store, client: client, modelKey: { model }, rrfK: rrfK)
    }

    /// Hybrid retrieval: cosine top-K on vectors + FTS5 BM25 top-K → RRF.
    ///
    /// When `sourceIds` is non-empty, both branches are restricted to chunks in
    /// those sources; an empty set (the default) means no filter.
    public func search(
        notebookId: Int64,
        query: String,
        topK: Int = 8,
        sourceIds: Set<Int64> = []
    ) async throws -> [RetrievalHit] {
        // 1) Vector ranking — embed query, score against stored vectors.
        let key = modelKey()
        let queryVectors = try await client.embed(model: key, inputs: [query])
        let queryVector = queryVectors.first ?? []
        let storeRef = store
        let modelRef = key
        let allEmbeddings = try await MainActor.run {
            try storeRef.embeddings(notebookId: notebookId, model: modelRef)
        }
        let candidateEmbeddings = sourceIds.isEmpty
            ? allEmbeddings
            : allEmbeddings.filter { sourceIds.contains($0.sourceId) }
        let scored: [(chunkId: Int64, sourceId: Int64, score: Float)] =
            candidateEmbeddings.map { e in
                (
                    chunkId: e.chunkId,
                    sourceId: e.sourceId,
                    score: Cosine.similarity(queryVector, e.vector.values)
                )
            }
        let vectorRanked: [(chunkId: Int64, sourceId: Int64, score: Float)] =
            Array(scored.sorted { $0.score > $1.score }.prefix(topK))

        // 2) FTS ranking — BM25 top-K on chunks_fts within the notebook.
        let queryRef = query
        let topKRef = topK
        let sourceIdsRef = sourceIds
        let ftsRanked = try await MainActor.run {
            try Self.ftsTopK(
                store: storeRef,
                notebookId: notebookId,
                query: queryRef,
                k: topKRef,
                sourceIds: sourceIdsRef
            )
        }

        // 3) Reciprocal Rank Fusion.
        var rrfScores: [Int64: Float] = [:]
        var meta: [Int64: (sourceId: Int64, snippet: String)] = [:]
        for (rank, hit) in vectorRanked.enumerated() {
            rrfScores[hit.chunkId, default: 0] += 1.0 / Float(rrfK + rank + 1)
            meta[hit.chunkId] = (hit.sourceId, "")
        }
        for (rank, hit) in ftsRanked.enumerated() {
            rrfScores[hit.chunkId, default: 0] += 1.0 / Float(rrfK + rank + 1)
            meta[hit.chunkId] = (hit.sourceId, hit.snippet)
        }

        // 4) Hydrate snippets for chunks that only came from the vector branch.
        let missingSnippets = meta.compactMap { (id, m) in m.snippet.isEmpty ? id : nil }
        if !missingSnippets.isEmpty {
            let snippets = try await MainActor.run {
                try Self.snippets(store: storeRef, chunkIds: missingSnippets)
            }
            for (id, snippet) in snippets {
                meta[id] = (meta[id]?.sourceId ?? 0, snippet)
            }
        }

        return rrfScores
            .sorted { $0.value > $1.value }
            .prefix(topK)
            .compactMap { (id, score) in
                guard let m = meta[id] else { return nil }
                return RetrievalHit(chunkId: id, sourceId: m.sourceId, score: score, snippet: m.snippet)
            }
    }

    // MARK: - Internal SQL helpers

    @MainActor
    private static func ftsTopK(
        store: NotebookStore,
        notebookId: Int64,
        query: String,
        k: Int,
        sourceIds: Set<Int64>
    ) throws -> [(chunkId: Int64, sourceId: Int64, snippet: String)] {
        try store.runOnDatabase { db in
            var arguments: [DatabaseValueConvertible] = [Self.escapeFTS(query), notebookId]
            var sourceFilter = ""
            if !sourceIds.isEmpty {
                let placeholders = sourceIds.map { _ in "?" }.joined(separator: ",")
                sourceFilter = "AND sc.source_id IN (\(placeholders))"
                arguments.append(contentsOf: sourceIds.map { $0 as DatabaseValueConvertible })
            }
            arguments.append(k)
            let rows = try Row.fetchAll(
                db,
                sql: """
                SELECT sc.id AS chunk_id, sc.source_id AS source_id, sc.text AS text
                FROM chunks_fts f
                JOIN source_chunks sc ON sc.id = f.chunk_id
                JOIN sources s ON s.id = sc.source_id
                WHERE f.text MATCH ? AND s.notebook_id = ? \(sourceFilter)
                ORDER BY bm25(chunks_fts)
                LIMIT ?
                """,
                arguments: StatementArguments(arguments)
            )
            return rows.map { r in
                let text: String = r["text"]
                return (
                    chunkId: r["chunk_id"],
                    sourceId: r["source_id"],
                    snippet: String(text.prefix(240))
                )
            }
        }
    }

    @MainActor
    private static func snippets(
        store: NotebookStore,
        chunkIds: [Int64]
    ) throws -> [Int64: String] {
        guard !chunkIds.isEmpty else { return [:] }
        return try store.runOnDatabase { db in
            let placeholders = chunkIds.map { _ in "?" }.joined(separator: ",")
            let rows = try Row.fetchAll(
                db,
                sql: "SELECT id, text FROM source_chunks WHERE id IN (\(placeholders))",
                arguments: StatementArguments(chunkIds)
            )
            var out: [Int64: String] = [:]
            for r in rows {
                let text: String = r["text"]
                out[r["id"]] = String(text.prefix(240))
            }
            return out
        }
    }

    /// Defensive escaping of double quotes for the FTS5 `MATCH` operator.
    /// We wrap the whole query in double quotes so it's treated as a phrase
    /// search and special characters can't break the parser.
    private static func escapeFTS(_ raw: String) -> String {
        let escaped = raw.replacingOccurrences(of: "\"", with: "\"\"")
        return "\"\(escaped)\""
    }
}
